pkg/deploy/registry/deploylog/rest.go

```diff
@@ -0,0 +1,181 @@
+	return c.WatchRcFn(ctx, label, field, rv)
+}
+
+type podGetter struct {
+	podsNamespacer kclient.PodsNamespacer
+}
+
+// Get is responsible for retrieving the deployer pod
+func (g *podGetter) Get(ctx kapi.Context, name string) (runtime.Object, error) {
+	ns, ok := kapi.NamespaceFrom(ctx)
+	if !ok {
+		return nil, errors.NewBadRequest("namespace parameter required.")
+	}
+	return g.podsNamespacer.Pods(ns).Get(name)
+}
+
+const defaultTimeout time.Duration = 10 * time.Second
+
+// NewREST creates a new REST for DeployLogs. It accepts a DeploymentClient and a pod
+// client to get necessary attributes to assemble the URL to which the request shall be
+// redirected in order to get the deployment logs.
+func NewREST(d DeploymentClient, pn kclient.PodsNamespacer, connectionInfo kclient.ConnectionInfoGetter) *REST {
+	return &REST{
+		Deploy:         d,
+		PodGetter:      &podGetter{pn},
+		ConnectionInfo: connectionInfo,
+		Timeout:        defaultTimeout,
+	}
+}
+
+// REST implements GetterWithOptions
+var _ = rest.GetterWithOptions(&REST{})
+
+// Get returns a streamer resource with the contents of the deployment log
+func (r *REST) Get(ctx kapi.Context, name string, opts runtime.Object) (runtime.Object, error) {
+	deployLogOpts, ok := opts.(*deployapi.DeploymentLogOptions)
+	if !ok {
+		return nil, errors.NewBadRequest("did not get an expected options.")
+	}
+
+	config, err := r.Deploy.GetDeploymentConfig(ctx, name)
+	if err != nil {
+		return nil, errors.NewNotFound("deploymentConfig", name)
+	}
+
+	deployment, err := r.Deploy.GetDeployment(ctx, deployutil.LatestDeploymentNameForConfig(config))
+	if err != nil {
+		// TODO: Fallback to latest-1?
+		return nil, errors.NewNotFound("deployment", deployutil.LatestDeploymentNameForConfig(config))
+	}
+
+	status := deployutil.DeploymentStatusFor(deployment)
+	switch status {
+	case deployapi.DeploymentStatusNew, deployapi.DeploymentStatusPending:
+		if deployLogOpts.NoWait {
+			glog.V(4).Infof("Deployment %s/%s is in %s state. No logs to retrieve yet.", deployment.Namespace, deployment.Name, status)
+			return &genericrest.LocationStreamer{}, nil
+		}
+		glog.V(4).Infof("Deployment %s/%s is in %s state, waiting for it to start...", deployment.Namespace, deployment.Name, status)
+		if err := r.waitForDeployment(ctx, deployment); err != nil {
+			return nil, err
+		}
+	}
+
+	deployPodName := deployutil.DeployerPodNameForDeployment(deployment.Name)
+	logOpts := &kapi.PodLogOptions{
+		Follow: deployLogOpts.Follow,
+	}
+	location, transport, err := pod.LogLocation(r.PodGetter, r.ConnectionInfo, ctx, deployPodName, logOpts)
+	if err != nil {
+		return nil, errors.NewBadRequest(err.Error())
+	}
+	return &genericrest.LocationStreamer{
+		Location:    location,
+		Transport:   transport,
+		ContentType: "text/plain",
+		Flush:       deployLogOpts.Follow,
+	}, nil
+}
+
+func (r *REST) waitForDeployment(ctx kapi.Context, deployment *kapi.ReplicationController) error {
```

Comment:
Of course, then the question is why can't watch.Interface provide that.
Why do we have to keep adding infrastructure on top of it, vs making the
client watch reliable?

I don't particularly care for this pull, but if I want reliable watch I
shouldn't have to add five different tools on top of watch.

On Jul 30, 2015, at 5:02 PM, Dan Mace <notifications@github.com> wrote:

In pkg/deploy/registry/deploylog/rest.go
<https://github.com/openshift/origin/pull/3943#discussion_r35920359>:

> +	logOpts := &kapi.PodLogOptions{
> +		Follow: deployLogOpts.Follow,
> +	}
> +	location, transport, err := pod.LogLocation(r.PodGetter, r.ConnectionInfo, ctx, deployPodName, logOpts)
> +	if err != nil {
> +		return nil, errors.NewBadRequest(err.Error())
> +	}
> +	return &genericrest.LocationStreamer{
> +		Location:    location,
> +		Transport:   transport,
> +		ContentType: "text/plain",
> +		Flush:       deployLogOpts.Follow,
> +	}, nil
> +}
> +
> +func (r *REST) waitForDeployment(ctx kapi.Context, deployment *kapi.ReplicationController) error {

Addendum: it's important in some cases to have a reliable *waiting*
mechanism, which is very annoying to get right with just watch.Interface,
hence the extra stuff we put around it.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/3943/files#r35920359>.